Emby.Server.Implementations/TV/TVSeriesManager.cs

```diff
@@ -189,40 +189,24 @@ private static string GetUniqueSeriesKey(Series series)
         /// <returns>Task{Episode}.</returns>
         private Tuple<DateTime, Func<Episode>> GetNextUp(string seriesKey, User user, DtoOptions dtoOptions)
         {
-            var lastWatchedEpisode = _libraryManager.GetItemList(new InternalItemsQuery(user)
+            Episode lastWatchedEpisode = null;
+            BaseItem[] episodes = null;
+            var series = _libraryManager.GetItemById(seriesKey) as Series;
+            if (series != null)
             {
-                AncestorWithPresentationUniqueKey = null,
-                SeriesPresentationUniqueKey = seriesKey,
-                IncludeItemTypes = new[] { nameof(Episode) },
-                OrderBy = new[] { new ValueTuple<string, SortOrder>(ItemSortBy.SortName, SortOrder.Descending) },
-                IsPlayed = true,
-                Limit = 1,
-                ParentIndexNumberNotEquals = 0,
-                DtoOptions = new DtoOptions
-                {
-                    Fields = new ItemFields[]
-                    {
-                        ItemFields.SortName
-                    },
-                    EnableImages = false
-                }
-            }).FirstOrDefault();
+                episodes = series.GetEpisodes(user, dtoOptions).ToArray();
+                lastWatchedEpisode = episodes.Reverse().FirstOrDefault(episode => episode.IsPlayed(user)) as Episode;
```

Comment:
Not 100% on this, but line 204 infers that episodes can be null. Just glancing at this from the browser, so I'm not sure if GetEpisodes can return null, but if it returns a null value to episodes, I believe episodes.Reverse().FirstOrDefault(...) would throw a NullReferenceException.

We can prevent this with null checking.

`lastWatchedEpisode = episodes?.Reverse().FirstOrDefault(episode => episode.IsPlayed(user)) as Episode;`

Reply:
GetEpisodes() cannot return null, the null check later is just for the case, that the series wasn't found and therefore GetEpisodes() was not executed.
Should I still check for null even though it is currently impossible?

Reply:
Apologize for the delay! I'm a huge fan of defensive programming, so I'll always recommend having a null check if there's ever a modicum of potential for a null to slip through. 

If there's ever an instance where a series is on your platform, but the episodes were removed for some reason, it might return null on series.GetEpisode().

Adding the ? check will verify that your application won't bomb out for any reason and can provide the flexibility of allowing a null episode list for expansion in the future

Reply:
Ok I wasn't sure if I should rely on the current implementation of GetEpisodes(). Now the null checks are in place.

I hope I understood your proccess correctly and I should squash the entire change to one commit, as the null checks are not a meaningful change for the final log

Reply:
Defensive programming in a statically typed language always sounds a bit like an excuse to be lazy imo. Reason about the code and its dependencies and act accordingly. Looking at `GetEpisodes` it doesn't look like it can ever return false, so adding extra null checks isn't necessary.

Reply:
I'm curious on your comment about defensive programming being lazy on statically typed languages. Do you mind expanding a bit on that?

As for the other context of your comments–I didn't actually pull the branch or anything. This was more of a quick glance at outstanding pull requests through the browser, so it's completely plausible that `GetEpisodes` can never ever return null, but I tried to express this in my original comment. 

I think it's better to have the null check here instead of not having it for future cases. Is there a time where you'd recommend not having it here?

Reply:
In the context of full control, defensive programming should not be required. It's either a sign of lazy developers or terrible design.

There's an implicit contract that `GetEpisodes` returns an IEnumerable, which may or may not be empty. If someone changes it to also return null, then that's a breach of said contract. Whoever is making any change has to ensure the continued correctness of the program.

Not carefully considering the impacts of the changes to existing code shows a lack of discipline and laziness imo. It should not be necessary to be programming defensively against your own peers!